lib/utils/compose.js

```diff
@@ -0,0 +1,1129 @@
+	// having to deal with splitting the string midway through some special
+	// character sequence.
+	return str.slice(0, str.lastIndexOf('\n'));
+};
+
+const LOG_LENGTH_MAX = 512 * 1024; // 512KB
+
+export function buildProject(
+	docker,
+	logger,
+	projectPath,
+	projectName,
+	composition,
+	arch,
+	deviceType,
+	emulated,
+	buildOpts,
+	inlineLogs,
+	convertEol,
+	dockerfilePath,
+) {
+	const _ = require('lodash');
+	const humanize = require('humanize');
+	const compose = require('resin-compose-parse');
+	const builder = require('resin-multibuild');
+	const transpose = require('docker-qemu-transpose');
+	const { BALENA_ENGINE_TMP_PATH } = require('../config');
+	const {
+		checkBuildSecretsRequirements,
+		makeBuildTasks,
+	} = require('./compose_ts');
+	const qemu = require('./qemu-ts');
+	const { toPosixPath } = builder.PathUtils;
+
+	logger.logInfo(`Building for ${arch}/${deviceType}`);
+
+	const imageDescriptors = compose.parse(composition);
+	const imageDescriptorsByServiceName = _.keyBy(
+		imageDescriptors,
+		'serviceName',
+	);
+
+	let renderer;
+	if (inlineLogs) {
+		renderer = new BuildProgressInline(
+			logger.streams['build'],
+			imageDescriptors,
+		);
+	} else {
+		const tty = require('./tty')(process.stdout);
+		renderer = new BuildProgressUI(tty, imageDescriptors);
+	}
+	renderer.start();
+
+	return Promise.resolve(checkBuildSecretsRequirements(docker, projectPath))
+		.then(() => qemu.installQemuIfNeeded(emulated, logger, arch, docker))
+		.tap(function(needsQemu) {
+			if (!needsQemu) {
+				return;
+			}
+			logger.logInfo('Emulation is enabled');
+			// Copy qemu into all build contexts
+			return Promise.map(imageDescriptors, function(d) {
+				if (typeof d.image === 'string' || d.image.context == null) {
+					return;
+				}
+				// external image
+				return qemu.copyQemu(path.join(projectPath, d.image.context), arch);
+			});
+		})
+		.then((
+			needsQemu, // Tar up the directory, ready for the build stream
+		) =>
+			tarDirectory(projectPath, { convertEol })
+				.then(tarStream =>
+					makeBuildTasks(
+						composition,
+						tarStream,
+						{ arch, deviceType },
+						logger,
+						projectName,
+					),
+				)
+				.map(function(/** @type {any} */ task) {
+					const d = imageDescriptorsByServiceName[task.serviceName];
+
+					// multibuild parses the composition internally so any tags we've
+					// set before are lost; re-assign them here
+					if (task.tag == null) {
+						task.tag = [projectName, task.serviceName].join('_').toLowerCase();
+					}
+					if (typeof d.image !== 'string' && d.image.context != null) {
+						d.image.tag = task.tag;
+					}
+
+					// configure build opts appropriately
+					if (task.dockerOpts == null) {
+						task.dockerOpts = {};
+					}
+					_.merge(task.dockerOpts, buildOpts, { t: task.tag });
+					if (typeof d.image !== 'string') {
+						/** @type {any} */
+						const context = d.image.context;
```

Comment:
`d.image.context` is typed as a string but we treat it as an object in the code, I don't know which is correct so I just keep the matching behavior from coffeescript to avoid breaking anything